File: practice/convert/example2_book_library/book-library/src/main/java/dev/starryeye/book_library/application/facade/ReturnBookFacade.java
package dev.starryeye.book_library.application.facade;

import dev.starryeye.book_library.application.book.query.GetBookService;
import dev.starryeye.book_library.application.facade.input.ReturnBookInput;
import dev.starryeye.book_library.application.user.command.ReturnBookService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@Component
@RequiredArgsConstructor
public class ReturnBookFacade {

    private final GetBookService getBookService;
    private final ReturnBookService returnBookService;

    public void returnBook(ReturnBookInput input) {

        if (!getBookService.existsBy(input.bookId())) {
            throw new IllegalArgumentException("book is not found, id = " + input.bookId());
        }

        returnBookService.returnBook(input.bookId(), input.userId());
    }
}

File: practice/convert/example2_book_library/kotlin/book-library/src/main/kotlin/dev/starryeye/book_library/application/facade/ReturnBookFacade.kt
package dev.starryeye.book_library.application.facade

import dev.starryeye.book_library.application.book.query.GetBookService
import dev.starryeye.book_library.application.facade.input.ReturnBookInput
import dev.starryeye.book_library.application.user.command.ReturnBookService
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Transactional
@Component
class ReturnBookFacade(
    private val getBookService: GetBookService,
    private val returnBookService: ReturnBookService,
) {

    fun returnBook(input: ReturnBookInput) {

        if (!getBookService.existsBy(input.bookId)) {
            throw IllegalArgumentException("book is not found, id = ${input.bookId}")
        }

        returnBookService.returnBook(userId = input.userId, bookId = input.bookId)
    }
}